P20-1212/Allow SafeMarkdown to format using typography

alex-m-brown opened the pull request:
SafeMarkdown currently wraps everything in a `<div>`, which can cause some issues with formatting. This PR adds compatibility with any component, including the Component Library's [Typography](https://github.com/code-dot-org/code-dot-org/blob/cb36ee4e9dce69f130134a427981b4def78a43f1/apps/src/componentLibrary/typography/Typography.tsx#L22) Components. 

 We have added an `unwrapped` property to return the markdown "unwrapped." This allows a `<SafeMarkdown/>` component to be wrapped in any other component. For example:
```
<Heading1 id="test-markdown-1">
  <SafeMarkdown unwrapped markdown="**Test** _markdown_!" />
</Heading1>
```
or
```
<p id="test-markdown-1">
  <SafeMarkdown unwrapped markdown="**Test** _markdown_!" />
</p>
```

## Links
Jira: [P20-1212](https://codedotorg.atlassian.net/browse/P20-1212)

## Testing story
Tests added to `SafeMarkdownTest.js`

carl-codeorg reviewed (approved):
LGTM

mgc1194 reviewed (approved):
I worked on these changes with Alex.
In order to support all types of [typography](https://github.com/code-dot-org/code-dot-org/tree/staging/apps/src/componentLibrary/typography) we decided to proceed with this implementation.
I'd like an additional review with an engineer with more React experience.

artem-vavilov commented:
Could you explain why you decided to move away from the first implementation?
https://github.com/code-dot-org/code-dot-org/pull/62964/commits/29b7c2991140a910e829ff3340c353cca92ef6c9

artem-vavilov commented:
Could you please explain why you decided to move away from the first implementation?
https://github.com/code-dot-org/code-dot-org/pull/62964/commits/29b7c2991140a910e829ff3340c353cca92ef6c9

alex-m-brown commented:
The first implementation, I misunderstood the ask for this ticket. I thought it meant native HTML elements, like `<p>`, and `<h1>`. After clarifying with Mario, the ask really was about using [Typography](https://github.com/code-dot-org/code-dot-org/blob/cb36ee4e9dce69f130134a427981b4def78a43f1/apps/src/componentLibrary/typography/Typography.tsx#L22) components, which added some extra layers of complexity, such as needing props and wanting to avoid duplicate wrapping.

artem-vavilov commented:
@alex-m-brown, would it be possible to extend the first implementation to support any components, such as `<p/>`, `<P/>`, or [any typography element](https://github.com/code-dot-org/code-dot-org/blob/P20-1212/Allow-SafeMarkdown-to-format-using-typography-/apps/src/componentLibrary/typography/TypographyElements.tsx) like `<BodyOneText/>`?

alex-m-brown commented:
Thanks to help from @artem-vavilov, SafeMarkdown should now take any component passed in via the `wrapperComponent` property.

artem-vavilov commented on apps/src/templates/SafeMarkdown.jsx:
```diff
@@ -39,27 +45,32 @@ class SafeMarkdown extends React.Component {
-    // returned) or we're assigning props. We prefer consistency over flexibility,
-    // so here we wrap the result in a div if it wasn't already
-    if (
-      rendered &&
-      rendered.type === 'div' &&
-      !Object.keys(markdownProps).length
```
I am worried that changing this condition might break something. I would recommend leaving it as it is:
```javascript
if (this.props.wrapperComponent) {
  const WrapperComponent = this.props.wrapperComponent;
  return (
    <WrapperComponent {...markdownProps}>
      {rendered.props.children}
    </WrapperComponent>
  );
}   
// rehype-react will only wrap the compiled markdown in a <div> tag
// if it needs to (ie, if there would otherwise be multiple elements
// returned) or we're assigning props. We prefer consistency over flexibility,
// so here we wrap the result in a div if it wasn't already
else if (
  rendered &&
  rendered.type === 'div' &&
  !Object.keys(markdownProps).length
) {
  return rendered;
} else {
  return <div {...markdownProps}>{rendered}</div>;
}
```

artem-vavilov reviewed (approved):
Great job! 🙌🏻

alex-m-brown commented:
Simplified this change greatly. Instead of passing in "wrapper" components to `SafeMarkdown`, we have added an `unwrapped` property to return the markdown "unwrapped." This allows a `<SafeMarkdown/>` component to be wrapped in any other component. For example:
```
<Heading1 id="test-markdown-1">
  <SafeMardown unwrapped markdown="**Test** _markdown_!" />
</Heading1>
```
or
```
<p id="test-markdown-1">
  <SafeMardown unwrapped markdown="**Test** _markdown_!" />
</p>
```

artem-vavilov commented on apps/test/unit/templates/SafeMarkdownTest.js:
```diff
@@ -60,6 +61,78 @@ describe('SafeMarkdown', () => {
+    expect(renderedHtml).toBe(
+      '<h1 id="test-markdown" class="heading-xxl"><strong>some</strong> <em>basic</em> <a href="markdown">inline</a></h1>'
+    );
+  });
```
I believe we can remove this test case, as it introduces an unnecessary dependency on another component without adding any new value.

artem-vavilov commented on apps/test/unit/templates/SafeMarkdownTest.js:
```diff
@@ -60,6 +61,78 @@ describe('SafeMarkdown', () => {
     ).toBe(true);
   });
 
+  it('will render markdown wrapped in an element', () => {
+    const paragraphWrapper = shallow(
+      <p id="test-markdown">
+        <SafeMarkdown
+          unwrapped
+          markdown={'**some** _basic_ [inline](markdown)'}
+        />
+      </p>
+    );
+
+    expect(paragraphWrapper.html()).toBe(
+      '<p id="test-markdown"><strong>some</strong> <em>basic</em> <a href="markdown">inline</a></p>'
+    );
```
I think this one test case should be more than enough to confirm that it returns the unwrapped string, so we can remove the others.

artem-vavilov reviewed (approved):
Clean and simple — perfect! Thank you, Alex!
